client/app/bundles/course/experience-points/components/ExperiencePointsTableRow.tsx

```diff
@@ -0,0 +1,196 @@
+import { FC, memo, useEffect, useState } from 'react';
+import { TableCell, TableRow } from '@mui/material';
+import equal from 'fast-deep-equal';
+import {
+  ExperiencePointsRecordMiniEntity,
+  ExperiencePointsRowData,
+} from 'types/course/experiencePointsRecords';
+
+import PointManagementButtons from 'course/users/components/buttons/PointManagementButtons';
+import NumberTextField from 'lib/components/core/fields/NumberTextField';
+import TextField from 'lib/components/core/fields/TextField';
+import Link from 'lib/components/core/Link';
+import { formatMiniDateTime } from 'lib/moment';
+
+interface Props {
+  id: number;
+  isStudentPage?: boolean;
+  record: ExperiencePointsRecordMiniEntity;
+  maxExp?: number;
+  isDownloading?: boolean;
+}
+
+const ExperiencePointsTableRow: FC<Props> = (props) => {
+  const { record, id, isStudentPage, maxExp, isDownloading } = props;
+  const [isDirty, setIsDirty] = useState(false);
+  const [errorHelperText, setErrorHelperText] = useState('');
+  const [errorReasonText, setErrorReasonText] = useState('');
+  const [rowData, setRowData] = useState({
+    id,
+    reason: '',
+    pointsAwarded: 0,
+  } as ExperiencePointsRowData);
+  const [defaultRowData, setDefaultRowData] = useState({
+    id,
+    reason: '',
+    pointsAwarded: 0,
+  } as ExperiencePointsRowData);
+  const isNotManuallyAwarded = maxExp;
+
+  useEffect(() => {
+    setRowData({
+      id: record.id,
+      reason: record.reason.text,
+      pointsAwarded: record.pointsAwarded,
+    });
+    setDefaultRowData({
+      id: record.id,
+      reason: record.reason.text,
+      pointsAwarded: record.pointsAwarded,
+    });
+  }, [record]);
+
+  const onUpdateReason = (value: string): void => {
+    const newData: ExperiencePointsRowData = { ...rowData, reason: value };
+    setIsDirty(value.trim().length > 0 && !equal(newData, defaultRowData));
+    setRowData(newData);
+  };
+
+  const onUpdatePoints = (value: string): void => {
+    const newData: ExperiencePointsRowData = {
+      ...rowData,
+      pointsAwarded:
+        Number.isNaN(Number(value)) || value === '' ? value : Number(value),
+    };
+
+    const defaultData: ExperiencePointsRowData = {
+      ...defaultRowData,
+      pointsAwarded: Number(defaultRowData.pointsAwarded),
+    };
+    setIsDirty(
+      !equal(newData, defaultData) && rowData.reason.trim().length > 0,
+    );
+    setRowData(newData);
+  };
+
+  const handleSave = (newData: ExperiencePointsRowData): void => {
+    if (!Number.isNaN(Number(newData.pointsAwarded))) {
+      const updateData: ExperiencePointsRowData = {
+        ...newData,
+        pointsAwarded: Number(newData.pointsAwarded),
+      };
+      setDefaultRowData({ ...updateData });
+      setIsDirty(false);
+    }
+  };
+
+  const renderReason = (): JSX.Element | string => {
```

Comment:
Create its own component for this: `ReasonCell` or something.

Same for your `<NumberTextField />` for the points cell.

That way, you dont have too many different usestate and logic inside this one component.